onnxruntime/core/providers/cpu/reduction/reduction_ops.h

```diff
@@ -132,6 +180,86 @@ class ReduceAggregatorSum : public ReduceAggregator<T, TVAL> {
   inline TVAL aggall(const T* from_data) {
     return Eigen::Map<const Eigen::Matrix<T, Eigen::Dynamic, 1>>(from_data, this->N_).sum();
   }
+
+  // Fast reduction
+  static inline FastReduceKind fast_reduce() { return FastReduceKindValues::KR | FastReduceKindValues::RK | FastReduceKindValues::KRK; }
+
+  static void FastReduceKR(const Tensor& input, const std::vector<int64_t>& fast_shape,
+                           Tensor& output, concurrency::ThreadPool* tp) {
+    ORT_ENFORCE(fast_shape.size() == 2, "Only works on matrices with two dimensions.");
+    ORT_ENFORCE(fast_shape[0] == output.Shape().Size(), "Output size mismatch.");
+    // TODO: use MLAS or BLAS
+    if (fast_shape[0] >= 4) {
+      const T* data = input.Data<T>();
```

Comment:
Please add comment explaining magic values like this so the next developer knows what the reason for branching here is.

Reply:
I'll fill cost information and will ThreadPool choose the best behvior.